loopy/schedule/checker/schedule.py

```diff
@@ -214,36 +341,411 @@ def generate_pairwise_schedules(
+
+    # {{{ _gather_blex_ordering_info(sync_kind): gather blex info for sync_kind
+
+    def _gather_blex_ordering_info(sync_kind):
+        """For the given sync_kind ("local" or "global"), create a mapping from
+        statement instances to blex space (dict), as well as a mapping
+        defining the blex ordering (isl map from blex space -> blex space)
+
+        Note that, unlike in the intra-thread case, there will be a single
+        blex ordering map defining the blex ordering for all statement pairs,
+        rather than separate (smaller) lex ordering maps for each pair
+        """
+
+        # {{{ First, create map from stmt instances to blex space.
+
+        # At the same time, gather information necessary to create the
+        # blex ordering map, i.e., for each loop, gather the 6 lex order tuples
+        # defined above in SpecialLexPointWRTLoop that will be required to
+        # create sub-maps which will be *excluded* (subtracted) from a standard
+        # lexicographic ordering in order to create the blex ordering
+
+        stmt_inst_to_blex = {}  # Map stmt instances to blex space
+        iname_to_blex_dim = {}  # Map from inames to corresponding blex space dim
+        blex_exclusion_info = {}  # Info for creating maps to exclude from blex order
+        blex_order_map_params = set()  # Params needed in blex order map
+        n_seq_blex_dims = 1  # Num dims representing sequential order in blex space
+        next_blex_tuple = [0]  # Next tuple of points in blex order
+
+        for lin_item in lin_items:
+            if isinstance(lin_item, EnterLoop):
+                enter_iname = lin_item.iname
+                if enter_iname in loops_with_barriers[sync_kind] - loops_to_ignore:
+                    pre_loop_blex_pt = next_blex_tuple[:]
+
+                    # Increment next_blex_tuple[-1] for statements in the section
+                    # of code between this EnterLoop and the matching LeaveLoop.
+                    next_blex_tuple[-1] += 1
+
+                    # Upon entering a loop, add one blex dimension for the loop
+                    # iteration, add second blex dim to enumerate sections of
+                    # code within new loop
+                    next_blex_tuple.append(enter_iname)
+                    next_blex_tuple.append(0)
+
+                    # Store 3 tuples that will be used later to create pairs
+                    # that will later be subtracted from the blex order map
+                    lbound = iname_bounds_pwaff[enter_iname][0]
+                    first_iter_blex_pt = next_blex_tuple[:]
+                    first_iter_blex_pt[-2] = lbound
+                    blex_exclusion_info[enter_iname] = {
+                        slex.PRE: tuple(pre_loop_blex_pt),
+                        slex.TOP: tuple(next_blex_tuple),
+                        slex.FIRST: tuple(first_iter_blex_pt),
+                        }
+                    # (make sure ^these are copies)
```

Comment:
? Could you explain?

Reply:
It's not important, and I'll remove it. It's just a not-so-obvious coding reminder for someone modifying the code in the future (me) that these three blex points, which are represented as lists before being stored in the dict as tuples, will be updated further as we go through the linearization, and that the original versions of these points (stored as lists) should not be stored directly in the dict without copying them first. Otherwise the blex points in the dict will be changed when the lists update.